Refine Hint Opaque/Transparent documentation

jfehrle opened the pull request:
Please take a look and let me know if I've gotten anything wrong.  Thanks.

@Janno?

jfehrle commented:
I'm not sure why I had to do a merge in github.  I synced master and tried to rebase this branch on it but git said it was up to date.

ppedrot commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -313,27 +313,56 @@ Creating hint databases
+
+Use :cmd:`Hint Opaque` and :cmd:`Hint Transparent` to control the opacity of
+individual items.  Hint opacity settings influence which hints the search tactics
+try, but they have no effect on how the selected tactic is executed.  The proof
```
Definitely not. Hint opacity is also used by the unification algorithm of the underlying apply-like tactics.

jfehrle replied:
"apply-like tactics" meaning Hint Resolve?

`apply` ignores regular opaque, so I don't see your point.  (And Hint Extern doesn't unfold goal or hint in any case.) 

```
Definition one := 1.
Opaque one.
Theorem thm : 1 = one. reflexivity. Qed.

Goal one = 1.
Succeed apply thm.  (* ignores opaque setting *)
Fail unfold one.
```

ppedrot replied:
apply-like tactics do rely on the transparent state even though they ignore it in some places.

jfehrle replied:
Is that only for tc eauto, which uses `autoapply` or are there other mechanisms?

Janno commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -504,28 +535,59 @@ Creating Hints
+
+            Create HintDb db.       (* constants, etc. transparent by default *)
+            Hint Opaque one : db.   (* except for "one" *)
+            Hint Resolve bar : db.  (* tactic is not tried if one is Hint Opaque *)
```
```suggestion
            Hint Resolve bar : db.  (* hint is not tried if one is Hint Opaque *)
```

Janno commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -722,11 +784,12 @@ Creating Hints
-   (first to last).  The groups are shown ordered alphabetically on the last component of
-   the symbol name.  Note that hints with the same cost are tried in
-   reverse of the order they're defined in, i.e., last to first.
+   This command displays all hints from database :n:`@ident`.  Hints are grouped by
+   the :term:`head constants <head constant>` of their patterns ("For ... ->").
+   The groups are shown ordered alphabetically on the last component of the head
+   constant name.   Within each group, hints are shown in the order in which they
+   will be tried (first to last).  Note that hints with the same cost are tried in
+   reverse of the order they're defined in, i.e., last defined is used first.
```
Could we mention hint modes here? I think they are also part of the output.

jfehrle replied:
Not yet.  I need to spend time playing with modes before I can say anything intelligent.

Janno commented on doc/sphinx/proofs/writing-proofs/equality.rst:
```diff
@@ -530,10 +530,8 @@ which reduction engine to use.  See :ref:`type-cast`.)  For example:
-       :tacn:`native_compute`
-       (see `#4476 <https://github.com/coq/coq/issues/4476>`_ and
-       :ref:`controlling-the-reduction-strategies`).
+       Opaque constants are not unfolded by most tactics
+       (see :ref:`controlling-the-reduction-strategies`).
```
```suggestion
       Constants marked opaque with :cmd:`Opaque` are not unfolded by most reduction tactics
       (see :ref:`controlling-the-reduction-strategies`).
```

Just making sure to distinguish constants made `Opaque` from truly opaque constants (defined with `Qed` or sealed using modules). And being more explicit about which tactics are affected. (Many tactics use unification which will happily ignore `Opaque`.)

jfehrle replied:
The Opaque command should mention this distinction.  Unification only pays attention to Qed opaqueness?  How can I create an example showing that?

How many tactics use unification?  Elsewhere I mention apply and reflexivity ignore Opaque.

How do you seal a constant with a module?  This is not sufficient:

![image](https://github.com/user-attachments/assets/a6ef5c0f-efac-4910-ae8d-b5fcf8a10e2c)

nor this:

![image](https://github.com/user-attachments/assets/c9ab0609-a68d-4190-a846-c46179da9525)

Janno replied:
> The `Opaque` command should mention this distinction. Unification only pays attention to Qed opaqueness? How can I create an example showing that?

```coq
Definition one : nat := 1.
Opaque one.
Theorem thm : 1 = one.
Proof.
  Succeed (lazymatch goal with |- ?l = ?r => unify l r end). (* tactic unification (most standard Coq tactics) *)
  Succeed refine (eq_refl).                                  (* evarconv unification (pretyper, ssreflect tactics) *)
  reflexivity.
Qed.
```

> How many tactics use unification? Elsewhere I mention apply and reflexivity ignore Opaque.

Almost everything uses unification in one way or another. My mental model is that `Opaque` only affects reduction tactics and anything else that somebody thought should honor `Opaque`. No example comes to mind, though.

> How do you seal a constant with a module? This is not sufficient:

You need module types for module sealing. I should have been more explicit. Sorry about that.

```coq
Module Type MT.
  Axiom one : nat.
End MT.
Module M : MT.
  Definition one : nat := 1.
End M.
Theorem thm : 1 = M.one. Fail reflexivity. Abort.
```

jfehrle replied:
Thanks.

Interestingly, dropping in `Transparent M.one` before the `Theorem` is executed but has no effect.  Also the `About` command doesn't report anything about the opacity status.

JasonGross replied:
> > Elsewhere I mention apply and reflexivity ignore Opaque.
> 
> Almost everything uses unification in one way or another. My mental model is that `Opaque` only affects reduction tactics and anything else that somebody thought should honor `Opaque`.

`Opaque` affects all conversion, including `reflexivity`, `apply`, `Qed`.  But it's treated differently (more strictly) by reduction tactics.  `Opaque` is roughly the same as `Strategy` infinity.  You can see this if you try to prove `fact 10 = 10 * fact 9` (or the symmetric version).  One of these will go instantaneously by `reflexivity`, unless you say `Opaque fact`, in which case it'll take a very long time.

jfehrle replied:
The `fact` expression you give takes .448 seconds vs 4.491 seconds for the symmetric expression, the latter much quicker than I expected.  Two thoughts:
- Should the section on conversion give this example and recommend setting computation-intensive items to `Opaque`?  Is it a simple thing everyone should do routinely?  Or can it be tricky to work with.
- Would there be high value in making `reflexivity` do something smarter?  I suspect that it maynot be simple even if it is conceptually possible.

> `Opaque` affects all conversion, including `reflexivity`, `apply`, `Qed`.

You may be right, but I wasn't to convince myself that it's true for `apply`.  Would be nice to elaborate/document this fully at some point.  Let's not do that now.

Janno commented on doc/sphinx/proofs/writing-proofs/equality.rst:
```diff
@@ -976,12 +974,17 @@ which supports additional fine-tuning.
 
-   Marks the specified constants as :term:`opaque` so tactics won't :term:`unfold` them
-   with :term:`delta-reduction`.
+   Marks the specified constants as :term:`opaque` so that most tactics won't
+   :term:`unfold` them with :term:`delta-reduction`.
```
```suggestion
   Marks the specified constants as :term:`opaque` so that most reduction tactics won't
   :term:`unfold` them with :term:`delta-reduction`.
```

jfehrle replied:
I've tended to refer to "conversion tactics", e.g. [here](https://coq.inria.fr/doc/V8.20.0/refman/proofs/writing-proofs/equality.html#applying-conversion-rules).  "reduction" seems less specific.

Janno reviewed (commented):
I am very happy to see more information in the docs! I left a bunch of comments.

jfehrle commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -313,27 +313,56 @@ Creating hint databases
+  pattern must match the goal before the hint will be tried.
+- If a hint other than `Hint Extern` contains unfoldable items, it will be tried
+  even if the pattern doesn't match.
+- `Extern` hints are tried if their pattern matches the goal while treating all
+  items in the hint pattern and the goal as opaque (i.e., a purely syntactic search)
```
I linked to wikipedia.  I hope they never change the link name :-)

Janno replied:
I am not sure linking to wikipeda is a good idea here. But I also cannot actually find the link (although that might be my browser's fault. Or github's.) There are many different notions of syntactic matching and even coq itself offers multiple interpretations, I think. If you mention Ltac `match` explicitly, you are giving users the necessary tool to debug their `Hint Extern` by copying the goal term and running `match goal_term with | hint_pattern => .. end`.

jfehrle commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -433,9 +462,10 @@ Creating Hints
-      subgoals generated by :tacn:`simple apply` :n:`@qualid` or, if specified, :n:`@natural`. The
+      The first form adds each :n:`@qualid` as a hint with the head symbol of the
+      type of :n:`@qualid` to the specified hint databases (:n:`@ident`\s). The
+      cost of the hint is the number of subgoals generated by :tacn:`simple apply`
+      :n:`@qualid` or, if specified, :n:`@natural`. The
```
I take whatever Print HintDb prints as accurate.  It is printing the actual tactic that will be executed.  So I think you're mistaken.  Feel free to prove me wrong.

"The cost of the hint is the number of subgoals generated" - that doesn't make much sense either because the costs are just constants in the hint and they can't depend on the result of applying the hint.  Unless tc eauto is actually adding up the costs.  I'll check that again.

ppedrot replied:
simple apply is different from what the internals actually use. As already mentioned, this is why we have autoapply. The debug printer is just that, a debug printer. Any similarity with reality would be fortuitous.

jfehrle replied:
Indeed.  We should disclose the difference in the tactics under `Print HintDb`.  @Janno if can you tell me what the best equivalents are, I will update `Hints.pr_hint`.  The 3 *auto tactics refer to slightly different code, so harder to tell if they are identical across the three.

How often do the best equivalents fail?  If it's more than a little, we should make the exact tactics available in Gallina.

Janno replied:
@ppedrot is more of an expert here as far as the implementation is concerned. I use `autoapply @hint with the_hint_db` to mimic the application of the hint. But AFAIK that is not exactly the tactic that is run by the OCaml code. I don't understand the differences, though.

jfehrle commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -313,27 +313,56 @@ Creating hint databases
+- For hints other than :cmd:`Hint Extern`, if the database is discriminated and
+  the pattern doesn't contain unfoldable (transparent) items, the full
+  pattern must match the goal before the hint will be tried.
+- If a hint other than `Hint Extern` contains unfoldable items, it will be tried
+  even if the pattern doesn't match.
```
I think this covers the first two bullets:

- For hints other than :cmd:`Hint Extern`, if the database is discriminated,
  all items that are opaque in both the goal and the pattern must match before the
  hint will be tried.

But not sure `discriminated` really matters.  I think the only difference for nondiscriminated is in the details of Hint Constructor, which I don't fully understand.  Is it possible and useful to have types such as `bool2` and `Hint Constructor bool2`?  Or to have arguments in such constructors?

```
Inductive bool2 : Prop :=
  | true2 : bool2
  | false2 : bool2.
Inductive True : Prop := (* for comparison *)
  I : True.

Hint Constructors bool2 : db.
Hint Constructors True : db.
Print HintDb db.

(*
For any goal ->   
For True ->   exact I (cost 0, pattern True, id 0)
For bool2 ->   exact false2 (cost 0, pattern bool2, id 0)
               exact true2 (cost 0, pattern bool2, id 0)
*)

  Goal true2.  (* The term "true2" has type "bool2" which should be Set, Prop or Type. *)
```

Janno replied:
I like your suggested formulation. Technically, `Hint Extern` is not an exception to the way you describe algorithm. In fact, with Coq 8.21, the patterns of `Hint Extern` are added to the dnet with full opacity. As a consequence, not even the syntactic pattern matching that was visible with `Set Typeclasses Debug` is shown now when the pattern does not match the goal. It would be fine to just remove the mention of `Hint Extern` here.

I am not familiar with `Hint Constructor` at all so I can't speak to that.

jfehrle commented:
@coqbot: run light build

jfehrle commented:
Updated.  @Janno, if you have no further comments, would you approve this?  IINM you're not authorized to merge PR.

@ppedrot Any further comments?

ppedrot commented on doc/sphinx/proofs/writing-proofs/equality.rst:
```diff
@@ -974,13 +972,28 @@ The commands to fine-tune the reduction strategies and the lazy conversion
+do unfolding is some cases, while :gdef:`Transparent <transparent>` permits
+unfolding.  Rocq has multiple notions of opaque:
+
+- **Permanently opaque.**  Theorems ending with :cmd:`Qed` are permanently
```
Don't introduce yet another terminology for this concept. We struggled to converge on something meaningful. You should pick the terminology used in #19029 already, otherwise the doc is going to end up being more confusing that it is today.

jfehrle replied:
Got it.

thomas-lamiaux commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -309,28 +358,71 @@ Creating hint databases
+  Otherwise (if not discriminated), all hints will be tried.  (Therefore, use
+  `discriminated` databases whenever possible.)
+- For `Extern` hints: If the hint has a pattern, hints matching the goal will be tried.
+  In this case, matching uses `syntactic unification
```
Unification in Coq is far more complicated than the wikipedia page. It is one of the most complicated thing in dependent proof assistants. I don't think it is good practice to link the wikipedia page as it is generic bachelor level ressource in CS that does not correspond to what happens in Coq.

jfehrle replied:
No doubt.  AFAIK there's no good description in the doc of how Rocq does unification that I could link to (IIUC there are 3 different unification algorithms).  Reverse engineering how unification works takes a lot of time; its documentation should go into its own future PR.

I think it's better to say something helpful even if it's incomplete.  Referring to syntactic unification is better than referring to "matching" with no further explanation.  How about wording like "matching uses an extended form of syntactic unification"?  Are you conversant with the differences?

Improving the documentation is an incremental process; lots of useful info is missing.  Adding the missing info usually takes considerable time to figure out what Rocq does from reading the code and creating examples that illustrate the differences.

jfehrle replied:
> I don't think it is good practice to link the wikipedia page as it is generic bachelor level ressource in CS

Rocq will not be successful if it only caters to PhDs.  As I've discussed with Théo, the documentation should be accessible to anyone with a bachelor's in computer science or math.  Rocq is not rocket science--the documentation is one of the main problems.

andres-erbsen replied:
Lacking a better resource, I think including the wikipedia link helps more than it hurts. Maybe "a form of syntactic unification" to hint to the reader that the rules from Wikipedia are not expected to transfer to Rocq exactly.

thomas-lamiaux commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -309,28 +358,71 @@ Creating hint databases
-   databases treat all constants as transparent, resulting in a larger
-   number of selected lemmas to be applied, and thus putting more pressure on
-   unification.
+   If there is no hint database named :n:`@ident`, creates a new hint database
+   with that name.  Otherwise, does nothing.
+
+   All constants, variables and projections are set to default to unfoldable (use
+   :cmd:`Hint Constants`, :cmd:`Hint Projections` and :cmd:`Hint Variables`
+   to change this).
+
+   By default, hint databases are undiscriminated.  We recommend using `discriminated`
+   because it generally performs better.
```
This intro is good.

thomas-lamiaux commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```````````````````````````diff
@@ -309,28 +358,71 @@ Creating hint databases
+Performance considerations
+``````````````````````````
+
+The proof search tactics decide which hints to try by first selecting hints
+whose pattern has the same :term:`head constant` as the goal.  Then the selected
+hints are tried as follows:
+
+- For hints other than :cmd:`Hint Extern`: If the database is discriminated, only
+  matching hints are tried.  Hints match if all compared items that are opaque in
+  both the goal and the pattern are identical.  In this case you can reduce the
+  number of matches by marking definitions as hint opaque through trial and error
+  (if the proof search fails after making something opaque, then it's incorrect).
+
+  Otherwise (if not discriminated), all hints will be tried.  (Therefore, use
+  `discriminated` databases whenever possible.)
+- For `Extern` hints: If the hint has a pattern, hints matching the goal will be tried.
+  In this case, matching uses `syntactic unification
```````````````````````````
This information is no longer in this paragraph 
> The database is
   implemented by a Discrimination Tree (DT) that serves as a filter to select
   the lemmas that will be applied.
   
What does discriminate means in this context.

thomas-lamiaux commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -431,16 +526,21 @@ Creating Hints
-      to the hint list.  If the type can be reduced to a type starting with a product,
-      :tacn:`simple apply` :n:`@qualid` is also added to the hints list.
+      of :n:`@qualid` does not start with a product, the command adds :tacn:`exact`
+      :n:`@qualid` to the hint list.  If the inferred type starts with a product,
+      the command adds :tacn:`simple apply` :n:`@qualid` to the hints list.  Note that
+      unlike :tacn:`auto` and :tacn:`eauto`,
```
Being reducable to a product and starting with a product is not the same. In once case, there is conversion / or hnf involved, in the other it is just a syntactic match. What is going on here ?

jfehrle replied:
This is determined when you add a hint (`Hint Resolve`), not when you're executing auto, et al.  For example:

```
Theorem x: forall a:nat,  a = a.
Admitted.
Theorem y: 1 = 2.
Admitted.

Create HintDb foo.
Hint Resolve x : foo.
Hint Resolve y : foo.
Print HintDb foo.

(*
Non-discriminated database
Unfoldable variable definitions: all
Unfoldable constant definitions: all
Unfoldable projection definitions: all
Cut: emp
For any goal ->   
For eq ->   exact y (level 0, pattern 1 = 2, id 0)
            simple apply x (level 0, pattern ?M151 = ?M151, id 0)
*)

Can you give an example of something that's reducible to a product?  We can try that, too, but I'm pretty sure `Hint Resolve` is doing any reduction.

thomas-lamiaux commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -431,16 +526,21 @@ Creating Hints
+      :n:`@qualid` to the hint list.  If the inferred type starts with a product,
+      the command adds :tacn:`simple apply` :n:`@qualid` to the hints list.  Note that
+      unlike :tacn:`auto` and :tacn:`eauto`,
+      :tacn:`typeclasses eauto` uses a tactic equivalent to :tacn:`autoapply`,
+      which behaves somewhat differently from :tacn:`simple apply`.
+      Nonetheless, that tactic's debug output and the hint database misleadingly show
+      :tacn:`simple apply`.
```
It should be in a warning

andres-erbsen replied:
Maybe a one-sentence warning along the lines of "Tactics printed in the debug output are similar to but not exactly the same as the ones executed by proof search" and then keep the paragraph mostly as is? Either way the change here is an improvement.

thomas-lamiaux reviewed (commented):
I was asked to look at it. I have added a few comments and one or two questions to try to clarify some text.

jfehrle commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -81,6 +85,48 @@ Tactics
       variant is very useful for getting a better understanding of automation, or
       to know what lemmas/assumptions were used.
 
+.. _info_auto_not_exact:
```
> From what I understand, this is specific to existential variable ? If so, it should be made explicit.

The doc doesn't describe what the constraint syntax does (below, from [here](![image](https://github.com/user-attachments/assets/2e0e1576-c068-44fb-9f37-cd685bb656cd)), so I was reluctant to be specific.  Better to be silent than to say something misleading.  Another research question.  Neither I nor, I believe, Pierre-Marie was able to construct a failing example using only existential variables that are not constraints.  Pierre-Marie provided the example after some discussion in a Coq Call.

![image](https://github.com/user-attachments/assets/5e50f259-894e-42de-ab6d-cbb57a5666a8)

>In any case, imo **turn the text and the example like the simple apply just above**, it will make the status of this remark clearer to readers.

What do you mean by the bold part?  I didn't understand.

jfehrle commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -309,28 +358,71 @@ Creating hint databases
-settings differ between these two methods of creation.  Databases created with
-:cmd:`Create HintDb` have the default setting `Transparent` for both `Variables`
-and `Constants`, while implicitly created databases have the `Opaque` setting.
+by adding a hint to an unknown database.  Databases created with
+:cmd:`Create HintDb` have the default setting `Transparent`
+for `Constants`, `Projections` and `Variables`, while implicitly created databases
+have the `Opaque` setting.  We recommend using :cmd:`Hint Constants`,
+:cmd:`Hint Projections` and :cmd:`Hint Variables` immediately after :cmd:`Create HintDb`
+to make these settings explicit.
```
Do you think this is really better?  It seems choppy to go to bullets for no obvious reason.  Is this really better than a paragraph?  (BTW, my system isn't generating bullets today, so I scribbled them on the screen shot.)

![image](https://github.com/user-attachments/assets/adacf96c-b5bf-4719-89b4-5734c8fa4fda)

jfehrle commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -309,28 +358,71 @@ Creating hint databases
+:cmd:`Hint Projections` and :cmd:`Hint Variables` immediately after :cmd:`Create HintDb`
+to make these settings explicit.
+
+Use :cmd:`Hint Opaque` and :cmd:`Hint Transparent` to control the opacity of
+individual items.  Hint opacity settings influence which hints the search tactics
+try, but they may or may not affect how the selected tactic is executed.
+(Usually not, except for :tacn:`typeclasses eauto`, which uses the equivalent of
+:tacn:`autoapply`, which explicitly uses the hint opacity settings when
+applying a :cmd:`Hint Resolve` hint.)
+The proof search tactics use unification to choose which tactics to try, for example
+whether the goal unifies with a theorem given by :cmd:`Hint Resolve`.
+These settings are distinct from the non-hint :cmd:`Opaque` and :cmd:`Transparent`
+settings.
```
FWIW "balise" is not an English word :-(.  Perhaps you mean a "warning box".  I think it is best used sparingly for extremely surprising behavior--yes, it draws attention, but it breaks up the logical flow of the text.

The `Hint Resolve` description includes:

![image](https://github.com/user-attachments/assets/1f0c575e-d8a2-446c-aa71-83575b52bcee)

and the `autoapply` description says:

![image](https://github.com/user-attachments/assets/556c3abc-405f-4764-936e-fa53c02e25e6)

Perhaps that is enough to cover this point without a warning box?

What do you think?

andres-erbsen commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -81,6 +85,48 @@ Tactics
+
+            Goal forall n, n=1 -> exists x y : nat, x = y /\ x = 0.
+            intros.
+            do 2 eexists; subst.      (* Fix 2: replace with "eexists; subst." twice *)
```
```suggestion
            do 2 (eexists; subst).      (* Fix 2: replace with "eexists; subst." twice *)
```

Is this the intended parse? Even if the rules are clear, I think it would be helpful to make it explicit here.

andres-erbsen commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -310,31 +359,76 @@ pattern or when the hint has no pattern.
+   when necessary.
+
+   By default, hint databases are undiscriminated.  We recommend using `discriminated`
+   because it generally performs better.
```
This also changes hint-picking behavior, right?

andres-erbsen commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -310,31 +359,76 @@ pattern or when the hint has no pattern.
+  matching hints are tried.  Hints match if all compared items that are opaque in
+  both the goal and the pattern are identical.  In this case you can reduce the
+  number of matches by marking definitions as hint opaque through trial and error
+  (if the proof search fails after making something opaque, then it's incorrect).
```
```suggestion
  both the goal and the pattern are identical. Marking definitions as Hint Opaque reduces the number of matches, but may cause some hints to no longer match the goal if they previously relied on unfolding of the marked definitions.
```

I don't think trial-and-error performance engineering on a hint database is recommendation-worthy -- for example, it isn't a good idea if there are users of that hintdb that cannot be tried out. Also, I worry that the resulting database would be brittle to small changes. I tried to capture the underlying principle in a more general comment.

jfehrle commented on doc/sphinx/proofs/automatic-tactics/auto.rst:
```diff
@@ -445,16 +542,22 @@ Creating Hints
-      to the hint list.  If the type can be reduced to a type starting with a product,
-      :tacn:`simple apply` :n:`@qualid` is also added to the hints list.
+      of :n:`@qualid` does not start with a product, the command adds :tacn:`exact`
+      :n:`@qualid` to the hint list.  If the type reduces to a type starting with a product,
```
Yes, it's an unnamed variation of `exact`.  As disclaimed elsewhere, the tactic names shown for proof search are not absolutely correct.

jfehrle commented on doc/sphinx/proofs/writing-proofs/equality.rst:
```diff
@@ -117,15 +117,21 @@ Rewriting with Leibniz and setoid equality
-   Replaces subterms with other subterms that have been proven to be equal.
-   The type of :n:`@one_term` must have the form:
+   Replaces subterms with other subterms that have been proven to be equal or logically
+   equivalent.  For equal terms, the type of :n:`@one_term` must have the form:
```
If it's a small change, IMO splitting it off is a waste of time and I'm likely to never return to this particular point.  This is documentation, not code.  The meaning of the change is obvious.  WDYT?

andres-erbsen commented:
This is a clear improvement and there's been plenty of time to discuss. Thank you, @jfehrle! Happy thanksgiving you too!

@coqbot merge now

andres-erbsen commented:
@coqbot run full ci

andres-erbsen commented:
@jfehrle The test failures look real. I sometimes find it easier to make a non-output test that through multiple calls checks the same thing instead of using debug output in a stable manner. If you confirm that what the CI founds still matches the understanding you documented here, I would be fine with splitting the tests into a separate PR so that this one can be merged.

```
FAILURES
    output/disc_uses_patterns.v ... Error! (unexpected output)
    output/disc_with_transparent.v ... Error! (unexpected output)
    output/extern_no_pattern.v ... Error! (unexpected output)
    output/nondisc_ignores_pattern.v ... Error! (unexpected output)
```

jfehrle commented:
@andres-erbsen I changed the output tests to success tests.  The Fail/Success commands were the important part.  The output was an aid to human understanding--and easy to recreate.  I squashed, please go ahead and merge.

Thanks for reviewing this!

I have another, shorter PR, #21021, that adds examples and some missing detail on the `destruct` tactic.  I submitted that on Aug 21 and it hasn't gotten any attention since late September.  If you're willing to look at that one, too, great, or I can try to find someone else to review it.
